### PR TITLE
 Don't eagerly call runMain when --start-paused is provided to web application

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Continuation of https://github.com/flutter/flutter/pull/39521. ~~This doesn't seem to work locally.~~ This now works

### CODE REVIEW

**Comment by DanTup** [comment]
Filed an issue at https://github.com/dart-lang/webdev/issues/716. I think the error may originate in the Chrome debugger - the resume signal is sent to it, but since it was never really paused, I think it rejects the `resume` call (since "not calling main" isn't exactly the same as pausing).

I don't know if there will need to be a fix here, but I think something will need to change in dwds first.

**Comment by DanTup** [comment]
@jonahwilliams I pulled the fix from https://github.com/dart-lang/webdev/issues/716 and initial testing seems like this works - in VS Code we successfully get the paused event, send breakpoints, then resume, and I can hit breakpoints and see the app load.

File: packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart [hunk 1]
```diff
@@ -232,7 +232,19 @@ class ResidentWebRunner extends ResidentRunner {
       unawaited(_vmService.registerService('reloadSources', 'FlutterTools'));
       websocketUri = Uri.parse(_connectionResult.debugConnection.uri);
       // Always run main after connecting because start paused doesn't work yet.
-      _connectionResult.appConnection.runMain();
+      if (!debuggingOptions.startPaused) {
+        _connectionResult.appConnection.runMain();
+      } else {
+        // I'm not sure if this is necessary.
```
Comment (DanTup) [inline packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart]: If we take this away, won't main never be called?

File: packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart [hunk 2]
```diff
@@ -232,7 +232,19 @@ class ResidentWebRunner extends ResidentRunner {
...
+        StreamSubscription<void> resumeSub;
+        resumeSub = _connectionResult.debugConnection.vmService.onDebugEvent.listen((vmservice.Event event) {
+          if (event.type == vmservice.EventKind.kResume) {
+            _connectionResult.appConnection.runMain();
+            resumeSub.cancel();
+          }
+        });
+        //
```
Comment (DanTup) [inline packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart]: nit: empty comment

**Review by DanTup — COMMENTED** [review]
Assuming the run-main-on-resume code is required, it'd be good to have a test for it (to prevent more "app hangs on white screen" bugs - we get a lot of them on mobile, for many different reasons).

Otherwise, is there anything preventing landing this? It would help me get more of Dart-Code's integration tests running on CI.

**Comment by jonahwilliams** [comment]
At this point I think I'm waiting for another release? Or was this actually working

**Comment by DanTup** [comment]
> At this point I think I'm waiting for another release?

Oh yeah 🙈

I was testing with `dependency_overrides` and forgot about that..

@grouma @jakemac53 I'm not familiar with how often you release dwds - is there likely to be one in the near future? I wouldn't make one specifically for this, but there have been a couple of significant fixes that'd help move VS Code along :-)

**Comment by jakemac53** [comment]
We typically do releases at least once a week right now given how quickly things are changing, and are planning on releasing today afaik (last release was last monday).

**Comment by DanTup** [comment]
Sounds good, thanks!

**Comment by devoncarew** [comment]
FYI, `0.7.5` of package:dwds is published now

File: packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart]: Yes, removed comment

File: packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart]: Fixed

**Comment by jonahwilliams** [comment]
Works on my computer!

**Review by DanTup — APPROVED** [review]
Great! 👍

**Pull request merged**